### ISSUE TITLE
Logging issues

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gix#74]
There were previously notes about logging issues in https://github.com/mridoni/gix/pull/25#issuecomment-1064165490 and #72.
As this should be tackled one way or the other I'm creating this tracking issue.

----------------------------

I'd suggest to _always_ use the logger for error handling and rename `GIXSQL_DEBUG_LOG ` to `GIXSQL_LOG` and `GIXSQL_DEBUG_LOG_ON` to `GIXSQL_DEBUG_LOGLEVEL`, defaulting to `ERR` (maybe have ´DBG` only be printed for `_DEBUG` - that would allow to toggle it to DBG during debugging, even if not compiled for that (or define the debug logging to nothing as it is done now).

Additional, if one compiles with `-D_DEBUG` (so logging is available in general) and logging in general is not activated by `GIXSQL_DEBUG_LOG_ON` - the error also goes to nirvana. With the log level I suggested errors would be default always be handled, if not _explicit_ turned OFF/0.
IF the log file is not explicit set it would either use a default one or stderr.

**Comment by GitMensch** [comment]
... and another thing at https://github.com/mridoni/gix/blob/8dd2cdbeaacf72c9300bccfe0c6ef23dfc09160d/gixsql/libgixsql/Logger.cpp#L88-L94

The logger checks - when not active - if it is active.
Each time, because that's a boolean.


The way to handle that would be an enum with the different levels and doing the initialization only once.
In the case of "there is no logger" this would mean something like

```c
	if(loglevel == LOG_OUTPUT_NOTSET){
		loglevel = com_get_loglevel();
		logfile = com_get_logfile();
	}
```

In the case of the Logger class that _may_ be done in its constructor.

This is what I've added to a local copy of ocesql to fix logging related issues:

```c
enum ocloglevel {
    LOG_OUTPUT_NOTSET = 0,
    LOG_OUTPUT_NOTHING = 1,
    LOG_OUTPUT_ERR = 2,
    LOG_OUTPUT_DEBUG = 3
};

static enum ocloglevel
env_to_loglevel (char *strenv)
{
	if (!strenv) {
		#if 0 /* better default to "see errors"... */
		return LOG_OUTPUT_NOTHING;
		#else
		return LOG_OUTPUT_ERR;
		#endif
	}
	if (!strcmp(strenv, "NOLOG") || !strcmp(strenv, "nolog")){
		return LOG_OUTPUT_NOTHING;
	}
	if (!strcmp(strenv, "ERR") || !strcmp(strenv, "err")){
		return LOG_OUTPUT_ERR;
	}
	if (!strcmp(strenv, "DBG") || !strcmp(strenv, "DEBUG") || !strcmp(strenv, "debug")){
		return LOG_OUTPUT_DEBUG;
	}
	{
		int ret = atoi(strenv);
		if (ret <= LOG_OUTPUT_NOTHING)  {
			return LOG_OUTPUT_NOTHING;
		}
		if (ret >= LOG_OUTPUT_DEBUG) {
			return LOG_OUTPUT_DEBUG;
		}
		return (enum ocloglevel)ret;
	}
}
```

**Comment by mridoni** [comment]
Yes, the logging mechanism should be almost completely overhauled, I admit it is quite hackish: it was only meant as a crutch to develop the runtime libraries, possibly with previously untested COBOL code; it worked (and works) well for that , but it's not usable as a user-level logging system.

**Comment by GitMensch** [comment]
Just an information:
I'm starting to do performance comparisons (against a bug-fixed, memory and and also logging-optimized ocesql), the first thing that showed up was actually in the logging part.
According to `perf record`, the function `Logger::check_debug_log()` took around 4% of the total runtime (because of the missing state as noted above; will re-run with logging disabled and create extra issues when I see something specific).

**Comment by mridoni** [comment]
Yes, I don't remember exact numbers, but I remember that in some production application that was heavily hitting the database, the debug feature was a performance killer, and that's why it is disabled in release builds.

**Comment by GitMensch** [comment]
When doing the logging overhaul it seems reasonable to:
* add the logging calls everywhere - but with a loglevel
* either let every class have its own static instance of the logger or create it as singleton (so one for all in the process, with the classes keeping a static reference to it; the later sounds more useful to me)
* wrap the logging calls into macros (one for each level)
* add a "NO_LOGGING" define that would redefine the actual logging _calls_ so only ERROR level calling is done in this case
* possibly: expose the current log level of the logger instance (either by an reference passed in via the constructor or via a public variable in it) and use that in the macros - this way a "logging" that is not activated costs only a single comparision against an enum, no function call invoked at all (that's more or less what I did to OCESQL)

```c
#define LOG(...)    {if (loglevel == LOG_OUTPUT_DEBUG) OCLOG(__FILE__, __FUNCTION__,__VA_ARGS__);}
#define ERRLOG(...) {if (loglevel != LOG_OUTPUT_NOTHING) OCERRLOG(__FILE__, __FUNCTION__,__VA_ARGS__);}
```
